#95 Improve prefix() unordered parallel performance for primitive streams

naftalmm opened the pull request.

loordgek commented:
1 claiming performance improvements without (jmh) benchmark should be considered meaningless

2 i cant see what you have changed

naftalmm commented:
> 1 claiming performance improvements without (jmh) benchmark should be considered meaningless

Fair point. Thanks for recommendation of performance testing framework.
Added appropriate benchmarks.

Got the following results:

Benchmark                                         (N)   Mode  Cnt     Score      Error  Units
PrefixDoubleStreamBenchmark.parallelOrdered    100000  thrpt   25  3844,307 ±  774,848  ops/s
PrefixDoubleStreamBenchmark.parallelUnordered  100000  thrpt   25  4358,190 ± 1166,045  ops/s
PrefixIntStreamBenchmark.parallelOrdered       100000  thrpt   25  2889,816 ±   40,818  ops/s
PrefixIntStreamBenchmark.parallelUnordered     100000  thrpt   25  4408,312 ±  131,077  ops/s
PrefixLongStreamBenchmark.parallelOrdered      100000  thrpt   25  2001,812 ±   29,942  ops/s
PrefixLongStreamBenchmark.parallelUnordered    100000  thrpt   25  4156,088 ±  106,256  ops/s

It seems that optimization for double stream is insignificant, so I've reverted those changes.

But optimizations for int & long showed almost x1.5 and x2 boost respectively.

> 2 i cant see what you have changed

I've done the same trick for primitive streams as it was done for object streams: I`ve written custom spliterators for unordered primitive streams and used them for constructing output stream in corresponding prefix() methods (if input stream had no ORDERED characteristic).

amaembo commented on src/main/java/one/util/streamex/PrefixOps.java:
```diff
@@ -1,302 +1,478 @@
+                return null;
+            }
+            if (accRef == null) {
+                accRef = new AtomicInteger();
```
Unfortunately, this implementation is incorrect and cannot be accepted. Here's a simple test that fails:
`assertEquals(1024, IntStreamEx.constant(2, 10).prefix((a, b) -> a*b).max().getAsInt());` (committed it)
It's not correct to feed the user-supplied operation with zero because it's not guaranteed that a zero would act as an identity element for any operation. OfUnordRef solves this using a sentinel, but it's not possible to have a sentinel inside AtomicInteger. It's really complex to make this correctly and faster than the current implementation. Likely, it's possible to do this with a StampedLock, but I have little experience with this.

naftalmm replied:
This was a major fail! Thanks for admitting it.
Managed to fix it keeping performance boost:

Benchmark                                       (N)   Mode  Cnt     Score     Error  Units
PrefixIntStreamBenchmark.parallelOrdered     100000  thrpt   25  2586,580 ±  98,957  ops/s
PrefixIntStreamBenchmark.parallelUnordered   100000  thrpt   25  4513,516 ± 408,134  ops/s
PrefixLongStreamBenchmark.parallelOrdered    100000  thrpt   25  2321,801 ±   5,534  ops/s
PrefixLongStreamBenchmark.parallelUnordered  100000  thrpt   25  3803,353 ± 628,167  ops/s

naftalmm replied:
Just realized I`ve introduced race condition with this fix.
Already repaired. Performance is still good:

Benchmark                                       (N)   Mode  Cnt     Score     Error  Units
PrefixIntStreamBenchmark.parallelOrdered     100000  thrpt   25  2646,411 ±  91,600  ops/s
PrefixIntStreamBenchmark.parallelUnordered   100000  thrpt   25  4571,769 ± 286,287  ops/s
PrefixLongStreamBenchmark.parallelOrdered    100000  thrpt   25  2325,392 ±  14,375  ops/s
PrefixLongStreamBenchmark.parallelUnordered  100000  thrpt   25  3882,683 ± 204,037  ops/s

amaembo replied:
What benchmark shows on your machine for a similar test for reference type? How many cores does your machine have? What are results for non-parallel execution?

naftalmm replied:
>How many cores does your machine have? 

8

>What are results for non-parallel execution?

PrefixIntStreamBenchmark.sequentialOrdered       100000  thrpt   25  1241,651 ± 17,490  ops/s
PrefixIntStreamBenchmark.sequentialUnordered     100000  thrpt   25  1585,448 ± 51,075  ops/s
PrefixLongStreamBenchmark.sequentialOrdered      100000  thrpt   25  1217,639 ± 20,058  ops/s
PrefixLongStreamBenchmark.sequentialUnordered    100000  thrpt   25  1390,811 ± 21,555  ops/s

Should I add these benchmarks to the repository too?

>What benchmark shows on your machine for a similar test for reference type?

PrefixObjectStreamBenchmark.parallelOrdered      100000  thrpt   25   319,191 ±  2,121  ops/s
PrefixObjectStreamBenchmark.parallelUnordered    100000  thrpt   25  1501,645 ± 37,583  ops/s
PrefixObjectStreamBenchmark.sequentialOrdered    100000  thrpt   25   814,598 ± 12,798  ops/s
PrefixObjectStreamBenchmark.sequentialUnordered  100000  thrpt   25   702,164 ±  7,094  ops/s

Should I add these benchmarks to the repository too?

naftalmm commented on .idea/streamex.iml:
```diff
@@ -12,5 +12,9 @@
     <orderEntry type="sourceFolder" forTests="false" />
     <orderEntry type="library" scope="TEST" name="Maven: junit:junit:4.12" level="project" />
     <orderEntry type="library" scope="TEST" name="Maven: org.hamcrest:hamcrest-core:1.3" level="project" />
+    <orderEntry type="library" name="Maven: org.openjdk.jmh:jmh-core:1.21" level="project" />
```
Moved them to separate IntelliJ module right in the repo root, but I belive it's not the same as a Maven module. Despite there is a POM, it has no connections(parent/module tags) with the main POM.
Is it ok?

amaembo replied:
For now, I believe, yes, it's ok.

amaembo commented:
GitHub UI displays huge diff in DoubleStreamEx and some other files, which complicates the review. Probably there's some bogus change like linebreaks? Is it possible to avoid this?

naftalmm commented on src/main/java/one/util/streamex/PrefixOps.java:
```diff
@@ -1,302 +1,518 @@
+        private static final class MyAtomicInteger extends AtomicInteger {
+            boolean init;
+            
+            public int getAndAccumulateOrInit(int x, IntBinaryOperator accumulatorFunction) {
```
>Having synchronization on every AtomicInteger update undermines the AtomicInteger usefulness at all.
M-m-m... I don't get this... AtomicInteger update inside the synchronized block happens exactly once per stream lifecycle (with set() method; its nessesary to make this operation synchronized with changing value of the init field).
After that getAndAccumulate() and accumulateAndGet() methods are called (size_of_stream-1) times in parallel manner fully utilizing AtomicInteger usefullness.

>Btw another possibility for OfUnordInt (but not for OfUnordLong) is to use AtomicLong instead and set initial value outside of the int range.
Interesting trick... Need to think about it...

naftalmm commented:
> GitHub UI displays huge diff in DoubleStreamEx and some other files, which complicates the review. Probably there's some bogus change like linebreaks? Is it possible to avoid this?

Yes. Git complained about CRLF line endings and I thought it happened that I've unintentionally introduced them, so I've manually changed them to LF.
But it seems files in repository were that way... Already reverted 'em back.
That's a pity GitHub UI is not that good as in IntelliJ IDEA (treating switch of line ending as one "meta-change", but not every line change):

![image](https://user-images.githubusercontent.com/10289217/84762883-0fba0c00-afd4-11ea-8f9a-4c5e141fbb2c.png)

>I'm worrying about decreased coverage.

Adding new tests revealed more bugs. Fixed 'em all. Perfomance speed-up is still present

Benchmark                                       (N)   Mode  Cnt     Score     Error  Units
PrefixIntStreamBenchmark.parallelOrdered     100000  thrpt   25  2511,918 ±  93,654  ops/s
PrefixIntStreamBenchmark.parallelUnordered   100000  thrpt   25  4231,807 ±  92,363  ops/s
PrefixLongStreamBenchmark.parallelOrdered    100000  thrpt   25  2311,495 ±   9,750  ops/s
PrefixLongStreamBenchmark.parallelUnordered  100000  thrpt   25  3843,591 ± 257,154  ops/s

naftalmm commented:
>Btw another possibility for OfUnordInt (but not for OfUnordLong) is to use AtomicLong instead and set initial value outside of the int range.

Tried this. Perfomance improved!

Benchmark                                      (N)   Mode  Cnt     Score     Error  Units
PrefixIntStreamBenchmark.parallelUnordered  100000  thrpt   25  5040,734 ± 121,767  ops/s

amaembo commented on src/main/java/one/util/streamex/PrefixOps.java:
```diff
@@ -299,4 +299,224 @@ public void accept(T next) {
+        
+        private void drain(LongConsumer action, long[] buf) {
+            long last = buf[idx - 1];
+            boolean accRefJustInitialized = accRef.initialize(last);
```
> M-m-m... I don't get this... AtomicInteger update inside the synchronized block happens exactly once per stream lifecycle

No it's not. The initialize synchronized method is called on the every drain that happens every time the buffer is full and with the same frequency as getAndAccumulate is called. It would probably easier to use `AtomicReference<AtomicLong>` and initialize AtomicReference in a CAS loop to avoid synchronization...

naftalmm replied:
> It would probably easier to use AtomicReference<AtomicLong> and initialize AtomicReference in a CAS loop to avoid synchronization...

Tried this, but performance degraded (from 3938,292 ± 106,900 down to 3373,426 ± 261,430  ops/s). Seems that wraping-unwraping longs to-from AtomicLong has significant overhead (or maybe my implementation of this idea was not good enough). Check it out:

```java
static final class OfUnordLong extends PrefixOps<Long, Spliterator.OfLong> implements LongConsumer, Spliterator.OfLong {
        private static final AtomicLong NONE = new AtomicLong();
        private final LongBinaryOperator localOp;
        private final BinaryOperator<AtomicLong> op;
        private boolean started;
        private AtomicReference<AtomicLong> accRef;
        private long acc;
        private LongPrefixBuffer buffer;
        
        OfUnordLong(Spliterator.OfLong source, LongBinaryOperator op) {
            super(source);
            this.localOp = op;
            //noinspection NumberEquality
            this.op = (a, b) -> a == NONE ? b : new AtomicLong(op.applyAsLong(a.get(), b.get()));
        }
    
        private final class LongPrefixBuffer extends PrefixBuffer implements LongConsumer {
            private final long[] buf = new long[BUF_SIZE];
            private long prevBufferLast;
        
            boolean init(Spliterator.OfLong source) {
                if (idx == 0) {
                    int i = 0;
                    while (i < BUF_SIZE && source.tryAdvance(this)) {
                        i++;
                    }
                    if (idx == 0) {
                        return false;
                    }
                
                    AtomicLong last = new AtomicLong(buf[idx - 1]);
                    isFirst = isFirst || accRef.compareAndSet(NONE, last);
                    if (!isFirst) {
                        prevBufferLast = accRef.getAndAccumulate(last, op).get();
                    }
                }
                return true;
            }
        
            void drainOne(LongConsumer action) {
                long value = buf[--idx];
                if (isFirst) {
                    action.accept(value);
                    if (idx == 0) {
                        isFirst = false;
                    }
                } else {
                    action.accept(localOp.applyAsLong(value, prevBufferLast));
                }
            }
        
            void drainAll(LongConsumer action) {
                if (!isInit()) return;
            
                if (isFirst) {
                    for (int i = 0; i < idx; i++) {
                        action.accept(buf[i]);
                    }
                    isFirst = false;
                } else {
                    for (int i = 0; i < idx; i++) {
                        action.accept(localOp.applyAsLong(buf[i], prevBufferLast));
                    }
                }
                idx = 0;
            }
        
            @Override
            public void accept(long value) {
                if (idx == 0) {
                    buf[idx++] = value;
                } else {
                    long prev = buf[idx - 1];
                    buf[idx++] = localOp.applyAsLong(prev, value);
                }
            }
        }
        
        @Override
        public Spliterator.OfLong trySplit() {
            if (started) {
                return null;
            }
            Spliterator.OfLong prefix = source.trySplit();
            if (prefix == null) {
                return null;
            }
            if (accRef == null) {
                accRef = new AtomicReference<>(NONE);
                buffer = new LongPrefixBuffer();
            }
            OfUnordLong pref = (OfUnordLong) doClone();
            pref.source = prefix;
            pref.buffer = new LongPrefixBuffer();
            return pref;
        }
        
        @Override
        public boolean tryAdvance(LongConsumer action) {
            if (accRef == null) {
                if (!source.tryAdvance(this)) {
                    return false;
                }
                action.accept(acc);
            } else {
                if (!buffer.init(source)) {
                    return false;
                }
                buffer.drainOne(action);
            }
            return true;
        }
        
        @Override
        public void forEachRemaining(LongConsumer action) {
            if (accRef == null) {
                source.forEachRemaining((LongConsumer) next -> {
                    if (started) {
                        acc = localOp.applyAsLong(acc, next);
                    } else {
                        acc = next;
                        started = true;
                    }
                    action.accept(acc);
                });
            } else {
                buffer.drainAll(action);
                long[] buf = new long[BUF_SIZE];
                source.forEachRemaining((LongConsumer) next -> {
                    if (idx == 0) {
                        buf[idx++] = next;
                    } else {
                        long prev = buf[idx - 1];
                        buf[idx++] = localOp.applyAsLong(prev, next);
                        if (idx == buf.length) {
                            drain(action, buf);
                            idx = 0;
                        }
                    }
                });
                if (idx > 0)
                    drain(action, buf);
            }
        }
        
        private void drain(LongConsumer action, long[] buf) {
            long last = buf[idx - 1];
            AtomicLong acc = accRef.getAndAccumulate(new AtomicLong(last), op);
            //noinspection NumberEquality
            if (acc != NONE) {
                for (int i = 0; i < idx; i++) {
                    action.accept(localOp.applyAsLong(buf[i], acc.get()));
                }
            } else {
                for (int i = 0; i < idx; i++) {
                    action.accept(buf[i]);
                }
            }
        }
        
        @Override
        public void accept(long next) {
            if (accRef == null) {
                if (started) {
                    acc = localOp.applyAsLong(acc, next);
                } else {
                    started = true;
                    acc = next;
                }
            } else {
                acc = accRef.accumulateAndGet(new AtomicLong(next), op).get();
            }
        }
    }
```

Also, tried to make use of StampedLock to avoid synchronized methods in MyAtomicLong but got no significant performance change, while code became much uglier:

```java
private static final class MyAtomicLongWithStampedLockInit extends AtomicLong {
        private final StampedLock sl = new StampedLock();
        private boolean init;
        
        public boolean initialize(long x) {
            if (!this.init) {
                boolean result = false;
                long stamp = sl.writeLock();
                if (!this.init) {
                    this.init = true;
                    set(x);
                    result = true;
                }
                sl.unlockWrite(stamp);
                return result;
            }
            if (sl.isWriteLocked()) {
                sl.unlockRead(sl.readLock());
            }
            return false;
        }
}
```
	
>The initialize synchronized method is called on the every drain that happens every time the buffer is full

Yes, it's true. My point was that the only operation on AtomicInteger inside this synchronized method happens exactly once per stream lifecycle. After that synchronized method become effectively equivalent to:

```java
public synchronized boolean initialize(long x) {
    if (!true) {}
    return false;
}
```

So it's still a bottleneck, but it seems it's not that narrow after all.

amaembo replied:
Well, my idea about double wrapping was slightly different. Like this (I have an unfinished draft):

```java
AtomicReference<AtomicLong> ref;
// in trySplit:
        if (ref == null) {
            ref = new AtomicReference<>();
        }

        private void drain(LongConsumer action, long[] buf) {
            long last = buf[idx - 1];
            AtomicLong sharedAcc = sharedAcc(last);
            if (sharedAcc == null) {
                for (int i = 0; i < idx; i++) {
                    action.accept(buf[i]);
                }
            }
            long acc = sharedAcc.getAndAccumulate(last, op); // op is just original op
            for (int i = 0; i < idx; i++) {
                action.accept(op.apply(buf[i], acc));
            }
        }

        private AtomicLong sharedAcc(long last) {
            AtomicLong sharedAcc = ref.get();
            if (sharedAcc == null && !ref.compareAndSet(null, new AtomicLong(last))) {
                sharedAcc = ref.get();
            }
            return sharedAcc;
        }
```

I haven't benchmarked it though. Probably it's no better.

naftalmm replied:
Wow, nice trick! It definetely better than my implementation of this idea, but unfortunately comparing to current implementation performance slightly degraded.

This is an isolated benchmark for different initialization strategies:

```java
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;
import java.util.concurrent.locks.StampedLock;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Level;
import org.openjdk.jmh.annotations.Param;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.infra.Blackhole;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.RunnerException;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

public class InitializeBenchmark {
    private static final class StampedLockInit extends AtomicLong {
        private final StampedLock sl = new StampedLock();
        private boolean init;
        
        public boolean initialize(long x) {
            if (!this.init) {
                boolean result = false;
                long stamp = sl.writeLock();
                if (!this.init) {
                    this.init = true;
                    set(x);
                    result = true;
                }
                sl.unlockWrite(stamp);
                return result;
            }
            if (sl.isWriteLocked()) {
                sl.unlockRead(sl.readLock());
            }
            return false;
        }
    }
    
    private static final class SynchronizedInit extends AtomicLong {
        private boolean init;
        
        /**
         * On the very first call sets the value to {@code x}
         *
         * @param x the initial value
         * @return {@code true} if it was the very first call
         */
        public synchronized boolean initialize(long x) {
            if (!init) {
                init = true;
                set(x);
                return true;
            }
            return false;
        }
    }
    
    private static final class AtomicReferenceInit {
        private final AtomicReference<AtomicLong> ref = new AtomicReference<>();
        /**
         * On the very first call sets the value to {@code x}
         *
         * @param x the initial value
         * @return {@code null} if it was the very first call
         */
        public AtomicLong getOrInitialize(long x) {
            AtomicLong sharedAcc = ref.get();
            if (sharedAcc == null && !ref.compareAndSet(null, new AtomicLong(x))) {
                sharedAcc = ref.get();
            }
            return sharedAcc;
        }
    }
    
    @State(Scope.Benchmark)
    public static class MyState {
        public ForkJoinPool fjp = new ForkJoinPool();
        private AtomicReferenceInit atomicReferenceInit;
        private StampedLockInit stampedLockInit;
        private SynchronizedInit synchronizedInit;
        private List<Callable<Boolean>> atomicReferenceInitTasks;
        private List<Callable<Boolean>> stampedLockInitTasks;
        private List<Callable<Boolean>> synchronizedInitTasks;
        
        @Param({"100000"})
        private int N;
    
        @Setup(Level.Trial)
        public void doSetup() {
            atomicReferenceInit = new AtomicReferenceInit();
            stampedLockInit = new StampedLockInit();
            synchronizedInit = new SynchronizedInit();
            atomicReferenceInitTasks = new ArrayList<>(N);
            stampedLockInitTasks = new ArrayList<>(N);
            synchronizedInitTasks = new ArrayList<>(N);
            for (int i = 0; i < N; i++) {
                long x = ThreadLocalRandom.current().nextLong();
                atomicReferenceInitTasks.add(() -> atomicReferenceInit.getOrInitialize(x) == null);
                stampedLockInitTasks.add(() -> stampedLockInit.initialize(x));
                synchronizedInitTasks.add(() -> synchronizedInit.initialize(x));
            }
        }
    }
    
    @Benchmark
    public void atomicReferenceInit(MyState state, Blackhole blackhole) {
        blackhole.consume(state.fjp.invokeAll(state.atomicReferenceInitTasks));
    }
    
    @Benchmark
    public void stampedLockInit(MyState state, Blackhole blackhole) {
        blackhole.consume(state.fjp.invokeAll(state.stampedLockInitTasks));
    }
    
    @Benchmark
    public void synchronizedInit(MyState state, Blackhole blackhole) {
        blackhole.consume(state.fjp.invokeAll(state.synchronizedInitTasks));
    }
    
    public static void main(String[] args) throws RunnerException {
        Options opt = new OptionsBuilder()
                .include(InitializeBenchmark.class.getSimpleName())
                .build();
        
        new Runner(opt).run();
    }
}
```


Got the following results on my computer:

```
Benchmark                                   (N)   Mode  Cnt   Score   Error  Units
InitializeBenchmark.atomicReferenceInit  100000  thrpt   25  47,046 ± 6,739  ops/s
InitializeBenchmark.stampedLockInit      100000  thrpt   25  48,853 ± 7,217  ops/s
InitializeBenchmark.synchronizedInit     100000  thrpt   25  57,980 ± 0,754  ops/s
```

amaembo commented:
Also, if you want to apply more changes, please do the following bookkeeping:
- Move benchmark sources from benchmark/src to benchmark/src/main/java
- Change benchmark package from prefix to one.util.streamex.benchmark.prefix

naftalmm commented on src/main/java/one/util/streamex/PrefixOps.java:
```diff
@@ -299,4 +299,224 @@ public void accept(T next) {
+        OfUnordInt(Spliterator.OfInt source, IntBinaryOperator op) {
+            super(source);
+            this.localOp = op;
+            this.op = (a, b) -> a == Long.MAX_VALUE ? b : op.applyAsInt((int) a, (int) b);;
```
>I'd prefer Math.toIntExact(a) instead of (int) a to be able to manifest a possible error.

I believe that kind of error is impossible here: all long operands here are basically int operands previously upcasted to long.

> Also, extra semicolon.
Fixed

naftalmm commented:
>The main thing that worries me is that current unordered implementation doesn't buffer the short-circuiting ops (via tryAdvance). This makes them contented and orders of magnitude slower than ordered short-circuiting ones.

Managed to fix this for parallel unordered streams (int, long & Object)

Benchmarks (with default jmh settings, ops/s) for optimized short-circuit ops (and several others for comparionship): 

Stream | Sequential | Parallel | Parallel (optimized)
------ | ---------- | -------- | ----------------------------------------------------------------
Int Ordered | 1567,396 ± 136,694 | 2605,660 ± 81,216
Int Unordered | 1536,674 ± 26,707 | 4399,797 ± 356,654
Int Ordered short-circuit | 1441,230 ± 84,124 | 1758,866 ± 123,153
Int Unordered short-circuit | 524,299 ± 19,462 | 80,377 ± 0,675 | 1317,539 ± 25,827
Long Ordered | 1198,769 ± 19,323 | 2283,106 ± 22,811
Long Unordered | 1370,030 ± 13,662 | 3938,292 ± 106,900
Long Ordered short-circuit | 1170,638 ± 15,687 | 1626,186 ±  53,911
Long Unordered short-circuit | 528,879 ±  9,324 | 89,650 ±   0,965 | 1261,431 ± 71,004
Object Ordered | 837,551 ± 2,087 | 313,451 ± 2,327
Object Unordered | 707,925 ± 2,923 | 1574,173 ± 27,857
Object Ordered short-circuit | 370,316 ± 23,002 | 296,136 ± 11,362
Object Unordered short-circuit | 340,078 ± 6,519 | 63,366 ± 1,302 | 950,797 ± 16,932

amaembo commented on src/test/java/one/util/streamex/TestHelpers.java:
```diff
@@ -182,6 +325,18 @@ static void repeat(int times, IntConsumer consumer) {
         }
     }
 
+    static void intStreamEx(Supplier<IntStream> base, Consumer<IntStreamExSupplier> consumer) {
```
I think it's easier to delegate to streamEx() without defining any additional types. Like `streamEx(() -> base.get().boxed(), s -> consumer.accept(s.get().mapToInt(x -> x)))`. Should be moreless similar but without much code duplication.

naftalmm commented:
The only lines that remain uncovered by tests in PrefixOps is the ones that would never be reached by any scenario of calls to StreamEx public API. This code was written just for exhauslively correct implementations of ```accept``` methods (for possible future usage), accounting all possible internal states of respectful spliterators, even the ones, in which that method would never be called (in current implementation).

To reach these lines the one should operate directly on OfUnord* spliterator and cast it to its exact type, something like that:
```java
    PrefixOps.OfUnordLong spliterator = (PrefixOps.OfUnordLong) LongStreamEx.constant(1, 5).unordered().prefix(Long::sum).spliterator();
    spliterator.trySplit();
    spliterator.accept(123);
    spliterator.accept(123);
```
Can't come up with any meaningful reason for user to do so.

Should I still cover these code lines with tests?

naftalmm commented:
On a second thought decided to remove them - they were primarely used in previous (not optimized) version of short-circuit operations and still don't add any sense for direct user call to ```accept``` methods
Besides those if-checks are not completely free and this PR is about performance.

amaembo commented:
Merged. Thanks for your work!
